### PR TITLE
Embedded Wasm

### PR DESCRIPTION
**Pull request by hjorthjort**
This is the first change to make KWasm embedded, i.e., not giving it an inherent way to load programs. This means any downstream projects will have to update their embedding accordingly.

The next step is to clean up the `test.md` file so that it doesn't do rewrites on the ~~`<wasmK>`~~`<instrs>`  cell, but rather passes control back and forth between its `<k>` cell and the ~~`<wasmK>`~~`<instrs>` cell. However, that change does not affect any (known) downstream projects.

NOTE: We are now importing `WASM-TEST` in `KWASM-LEMMAS`. This is not ideal, because there should be no reason to import the testing embedding. However, due to the issues [here](https://github.com/kframework/k/issues/1347) we can't get the tests to run without it. We should change back to importing just `WASM-TEXT` when the upstream issues are fixed.

### CODE REVIEW

File: test.md [hunk 1]
```diff
@@ -19,14 +19,33 @@ module WASM-TEST
     imports WASM-TEXT
 ```
 
+Configuration
+-------------
+
+```k
+    configuration
+      <k> $PGM:Stmts </k>
+      <wasm/>
```
Comment (ehildenb) [inline test.md:28]: We should always have a surrounding configuration for grabbing onto, so do `<wasm-test> <k> ... </k> <wasm/> </wasm-test>`

**Comment by ehildenb** [comment]
Post performance numbers here for `master` branch and this change please. I suspect this may have a serious negative impart on performance.

File: test.md (hunk 1 above)
Comment (hjorthjort) [reply test.md:28]: Done

**Comment by hjorthjort** [comment]
Performance with this change:
```
make test
1585.53user 51.00system 8:52.74elapsed 307%CPU (0avgtext+0avgdata 2873988maxresident)k
59280inputs+950864outputs (31major+15950391minor)pagefaults 0swaps

make test-conformance
16039.11user 136.53system 26:15.85elapsed 1026%CPU (0avgtext+0avgdata 9170876maxresident)k
8inputs+808640outputs (0major+45785105minor)pagefaults 0swaps

make test-simple  693.02s user 20.00s system 360% cpu 3:17.73 total
make test-prove   620.77s user 23.37s system 246% cpu 4:21.34 total
```

# `master`
```
make test:
1461.04user 49.38system 7:41.52elapsed 327%CPU (0avgtext+0avgdata 3649228maxresident)k
1272inputs+641328outputs (1major+15756522minor)pagefaults 0swaps

make test-prove:
16669.10user 136.45system 26:02.51elapsed 1075%CPU (0avgtext+0avgdata 8877116maxresident)k
4592inputs+803264outputs (1major+45695884minor)pagefaults 0swaps

make test-simple  649.49s user 19.23s system 353% cpu 3:09.12 total  
make test-prove   497.13s user 19.63s system 236% cpu 3:38.68 total
```

**Comment by ehildenb** [comment]
I assume the `test-prove` in the second block on `master` is supposed to be `test-conformance`?

File: tests/proofs/wrc20-spec.k [hunk 2]
```diff
@@ -7,7 +7,7 @@ module WRC20-SPEC
 
     // Reverse bytes spec.
 
-    rule <k> sequenceDefns(#t2aDefns<#freshCtx()>(#wrc20ReverseBytes)) // TODO: Have this pre-loaded in the store.
+    rule <instrs> sequenceDefns(#t2aDefns<#freshCtx()>(#wrc20ReverseBytes)) // TODO: Have this pre-loaded in the store.
```
Comment (ehildenb) [inline tests/proofs/wrc20-spec.k:10]: Formatting, line up `sequenceInstrs` with `sequenceDefns`, etc...

File: wasm.md [hunk 3]
```diff
@@ -181,25 +181,25 @@ The sorts `EmptyStmt` and `EmptyStmts` are administrative so that the empty list
...
-When a single value ends up on the instruction stack (the `<k>` cell), it is moved over to the value stack (the `<valstack>` cell).
+When a single value ends up on the instruction stack (the `<instrs>` cell), it is moved over to the value stack (the `<valstack>` cell).
 If the value is the special `undefined`, then `trap` is generated instead.
 
 ```k
-    rule <k> undefined => trap ... </k>
-    rule <k>        V:Val    => .        ... </k>
+    rule <instrs> undefined => trap ... </instrs>
+    rule <instrs>        V:Val    => .        ... </instrs>
          <valstack> VALSTACK => V : VALSTACK </valstack>
```
Comment (ehildenb) [inline wasm.md:203]: Formatting

File: wasm.md [hunk 4]
```diff
@@ -1471,7 +1471,7 @@ A new module instance gets allocated.
 Then, the surrounding `module` tag is discarded, and the definitions are executed, putting them into the module currently being defined.
 
 ```k
-    rule <k> #module(... types: TS, funcs: FS, tables: TABS, mems: MS, globals: GS, elem: EL, data: DAT, start: S,  importDefns: IS, exports: ES,
+    rule <instrs> #module(... types: TS, funcs: FS, tables: TABS, mems: MS, globals: GS, elem: EL, data: DAT, start: S,  importDefns: IS, exports: ES,
                          metadata: #meta(... id: OID, funcIds: FIDS))
           => sequenceDefns(TS)
           ~> sequenceDefns(IS)
```
Comment (ehildenb) [inline wasm.md:1477]: Formatting, line up the right side of `=>` and `~>` and `...` with the right side of `<instrs>`.

File: wasm.md [hunk 5]
```diff
@@ -1031,18 +1031,18 @@ Sort `Signedness` is defined in module `BYTES`.
...
+    rule <instrs> ITYPE . LOP:LoadOp MEMARG:MemArg => ITYPE . LOP (IDX +Int #getOffset(MEMARG)) ... </instrs>
          <valstack> < i32 > IDX : VALSTACK    => VALSTACK </valstack>
 
-    rule <k> load { ITYPE WIDTH EA SIGN }
+    rule <instrs> load { ITYPE WIDTH EA SIGN }
           => < ITYPE > #if SIGN ==K Signed
                            #then #signedWidth(WIDTH, #getRange(DATA, EA, WIDTH))
                            #else #getRange(DATA, EA, WIDTH)
                        #fi
          ...
```
Comment (ehildenb) [inline wasm.md:1044]: Formatting, line up `=>` and `...` and `<instrs>` on the right side.

File: wasm.md [hunk 6]
```diff
@@ -750,12 +750,12 @@ The `#take` function will return the parameter stack in the reversed order, then
...
     syntax Instr ::= "(" "invoke" Int ")"
  // -------------------------------------
-    rule <k> ( invoke FADDR )
+    rule <instrs> ( invoke FADDR )
           => init_locals #revs(#take(lengthValTypes(TDOMAIN), VALSTACK)) ++ #zero(TLOCALS)
           ~> block [TRANGE] INSTRS end
           ~> frame MODIDX TRANGE #drop(lengthValTypes(TDOMAIN), VALSTACK) LOCAL DEPTH IDS
           ...
-          </k>
+          </instrs>
```
Comment (ehildenb) [inline wasm.md:758]: Formatting.

File: wasm.md [hunk 7]
```diff
@@ -428,17 +428,17 @@ The various `init_local` variants assist in setting up the `locals` cell.
...
+    rule <instrs> init_locals VALUES => #init_locals 0 VALUES ... </instrs>
 
-    rule <k> #init_locals _ .ValStack => . ... </k>
-    rule <k> #init_locals N (VALUE : VALSTACK)
+    rule <instrs> #init_locals _ .ValStack => . ... </instrs>
+    rule <instrs> #init_locals N (VALUE : VALSTACK)
           => init_local N VALUE
           ~> #init_locals (N +Int 1) VALSTACK
           ...
-          </k>
```
Comment (ehildenb) [inline wasm.md:441]: Formatting.

File: tests/proofs/loops-spec.k [hunk 8]
```diff
@@ -4,7 +4,7 @@ module LOOPS-SPEC
     imports KWASM-LEMMAS
 
     // Lemma
-    rule <k> br 0
+    rule <instrs> br 0
              ~> label // Loop label.
                   [ .ValTypes ]
                   { loop .TypeDecls
```
Comment (ehildenb) [inline tests/proofs/loops-spec.k:10]: formatting

File: tests/proofs/functions-spec.k [hunk 9]
```diff
@@ -14,51 +14,51 @@ endmodule
...
+    rule <instrs> run  ( #getRange ( #setRange ( #setRange ( #setRange ( #setRange ( #setRange ( #setRange ( #setRange ( #setRange ( BM , ADDR , #getRange ( BM , ADDR +Int 7 , 1 ) , 1 ) , ADDR +Int 1 , #getRange ( BM , ADDR +Int 6 , 1 ) , 1 ) , ADDR +Int 2 , #getRange ( BM , ADDR +Int 5 , 1 ) , 1 ) , ADDR +Int 3 , #getRange ( BM , ADDR +Int 4 , 1 ) , 1 ) , ADDR +Int 4 , #getRange ( BM , ADDR +Int 3 , 1 ) , 1 ) , ADDR +Int 5 , #getRange ( BM , ADDR +Int 2 , 1 ) , 1 ) , ADDR +Int 6 , #getRange ( BM , ADDR +Int 1 , 1 ) , 1 ) , ADDR +Int 7 , #getRange ( BM , ADDR , 1 ) , 1 ) , ADDR +Int 3 , 1 ) )
           => done ( #getRange ( BM , ADDR +Int 4 , 1 ) )
          ...
```
Comment (ehildenb) [inline tests/proofs/functions-spec.k:61]: formatting

File: tests/proofs/functions-spec.k [hunk 10]
```diff
@@ -14,51 +14,51 @@ endmodule
...
+    rule <instrs> run  ( #setRange ( BM , ADDR modInt 4294967296 , #getRange ( BM , ADDR modInt 4294967296 , #numBytes ( ITYPE ) ) , #numBytes ( ITYPE ) ) )
           => done ( BM )
          ...
-         </k>
+         </instrs>
 
-    rule <k> run  ( (#getRange(BM, ADDR +Int 1, 1) +Int (#getRange (BM, ADDR, 1) <<Int 8)) modInt 65536 )
+    rule <instrs> run  ( (#getRange(BM, ADDR +Int 1, 1) +Int (#getRange (BM, ADDR, 1) <<Int 8)) modInt 65536 )
           => done (  #getRange(BM, ADDR +Int 1, 1) +Int (#getRange (BM, ADDR, 1) <<Int 8) )
          ...
```
Comment (ehildenb) [inline tests/proofs/functions-spec.k:52]: Formatting

File: tests/proofs/functions-spec.k [hunk 11]
```diff
@@ -14,51 +14,51 @@ endmodule
...
           => done ( #setRange(#setRange(BM, ADDR, #getRange(BM, ADDR +Int 1, 1), 1), ADDR +Int 1, #getRange(BM, ADDR, 1), 1) )
          ...
-         </k>
+         </instrs>
       requires ADDR >=Int 0
 
-    rule <k> run  ( #setRange ( BM , ADDR modInt 4294967296 , #getRange ( BM , ADDR modInt 4294967296 , #numBytes ( ITYPE ) ) , #numBytes ( ITYPE ) ) )
+    rule <instrs> run  ( #setRange ( BM , ADDR modInt 4294967296 , #getRange ( BM , ADDR modInt 4294967296 , #numBytes ( ITYPE ) ) , #numBytes ( ITYPE ) ) )
           => done ( BM )
          ...
```
Comment (ehildenb) [inline tests/proofs/functions-spec.k:47]: Formatting

File: tests/proofs/functions-spec.k [hunk 12]
```diff
@@ -14,51 +14,51 @@ endmodule
...
           => done ( ( ( #getRange ( BM , ADDR +Int 1 , 1 )            ) +Int ( #getRange ( BM , ADDR , 1 )            <<Int 8 ) ) <<Int 48 )
          ...
-         </k>
+         </instrs>
 
-    rule <k> run  ( #setRange(BM, ADDR, #getRange(BM, ADDR +Int 1, 1) +Int (#getRange(BM, ADDR, 1) <<Int 8), 2) )
+    rule <instrs> run  ( #setRange(BM, ADDR, #getRange(BM, ADDR +Int 1, 1) +Int (#getRange(BM, ADDR, 1) <<Int 8), 2) )
           => done ( #setRange(#setRange(BM, ADDR, #getRange(BM, ADDR +Int 1, 1), 1), ADDR +Int 1, #getRange(BM, ADDR, 1), 1) )
          ...
-         </k>
```
Comment (ehildenb) [inline tests/proofs/functions-spec.k:42]: Formatting

File: tests/proofs/functions-spec.k [hunk 13]
```diff
@@ -14,51 +14,51 @@ endmodule
...
+         </instrs>
 
-    rule <k> run  ( ( ( #getRange ( BM , ADDR +Int 1 , 7 ) modInt 256 ) +Int ( #getRange ( BM , ADDR , 8 ) modInt 256 <<Int 8 ) ) <<Int 48 )
+    rule <instrs> run  ( ( ( #getRange ( BM , ADDR +Int 1 , 7 ) modInt 256 ) +Int ( #getRange ( BM , ADDR , 8 ) modInt 256 <<Int 8 ) ) <<Int 48 )
           => done ( ( ( #getRange ( BM , ADDR +Int 1 , 1 )            ) +Int ( #getRange ( BM , ADDR , 1 )            <<Int 8 ) ) <<Int 48 )
          ...
-         </k>
+         </instrs>
```
Comment (ehildenb) [inline tests/proofs/functions-spec.k:37]: Formatting

File: tests/proofs/functions-spec.k [hunk 14]
```diff
@@ -14,51 +14,51 @@ endmodule
...
+    rule <instrs> run ( #getRange(BM, ADDR, 3) modInt (2 ^Int (8 *Int 3)) ) => done ( #getRange(BM, ADDR       , 3) ) ... </instrs>
+    rule <instrs> run ( #getRange(BM, ADDR, 7) modInt (2 ^Int (8 *Int 3)) ) => done ( #getRange(BM, ADDR       , 3) ) ... </instrs>
+    rule <instrs> run ( #getRange(BM, ADDR, 2) >>Int 8                    ) => done ( #getRange(BM, ADDR +Int 1, 1) ) ... </instrs> requires 0 <=Int ADDR
 
-    rule <k> run  ( ( ( #getRange ( BM , ADDR +Int 1 , 1 ) modInt 256 ) +Int ( #getRange ( BM , ADDR , 1 ) modInt 256 <<Int 8 ) ) <<Int 48 )
+    rule <instrs> run  ( ( ( #getRange ( BM , ADDR +Int 1 , 1 ) modInt 256 ) +Int ( #getRange ( BM , ADDR , 1 ) modInt 256 <<Int 8 ) ) <<Int 48 )
           => done ( (   #getRange ( BM , ADDR +Int 1 , 1 )              +Int ( #getRange ( BM , ADDR , 1 )            <<Int 8 ) ) <<Int 48 )
          ...
-         </k>
+         </instrs>
```
Comment (ehildenb) [inline tests/proofs/functions-spec.k:32]: Formatting

File: test.md [hunk 15]
```diff
@@ -476,7 +497,7 @@ The modules are cleaned all together after the test file is executed.
 ```k
     syntax Auxil ::= "#clearConfig"
  // -------------------------------
-    rule <k>    #clearConfig => . ...     </k>
+    rule <instrs>    #clearConfig => . ...     </instrs>
          <curModIdx>       _ => .Int      </curModIdx>
          <valstack>        _ => .ValStack </valstack>
          <locals>          _ => .Map      </locals>
```
Comment (ehildenb) [inline test.md:503]: Formatting

File: wasm.md (hunk 3 above)
Comment (hjorthjort) [reply wasm.md]: Done

File: wasm.md (hunk 4 above)
Comment (hjorthjort) [reply wasm.md]: Done

File: wasm.md (hunk 6 above)
Comment (hjorthjort) [reply wasm.md]: Done

File: test.md (hunk 15 above)
Comment (hjorthjort) [reply test.md]: Done

File: tests/proofs/functions-spec.k (hunk 14 above)
Comment (hjorthjort) [reply tests/proofs/functions-spec.k]: Done

File: tests/proofs/functions-spec.k (hunk 13 above)
Comment (hjorthjort) [reply tests/proofs/functions-spec.k]: Done

File: tests/proofs/functions-spec.k (hunk 12 above)
Comment (hjorthjort) [reply tests/proofs/functions-spec.k]: Done

File: tests/proofs/functions-spec.k (hunk 11 above)
Comment (hjorthjort) [reply tests/proofs/functions-spec.k]: Done

File: tests/proofs/functions-spec.k (hunk 10 above)
Comment (hjorthjort) [reply tests/proofs/functions-spec.k]: Done

File: tests/proofs/functions-spec.k (hunk 9 above)
Comment (hjorthjort) [reply tests/proofs/functions-spec.k]: Done

File: wasm.md (hunk 5 above)
Comment (hjorthjort) [reply wasm.md]: Done

File: tests/proofs/wrc20-spec.k (hunk 2 above)
Comment (hjorthjort) [reply tests/proofs/wrc20-spec.k:10]: Done

File: wasm.md (hunk 7 above)
Comment (hjorthjort) [reply wasm.md]: Done

File: tests/proofs/loops-spec.k (hunk 8 above)
Comment (hjorthjort) [reply tests/proofs/loops-spec.k]: Done

**Pull request merged**